### PR TITLE
Adds onboarding feature on the home page GLVSC-613

### PR DESCRIPTION
**Pull request by nzaytsev**
# Description

Shows progress tracker and onboarding items according to UsageTracker data

Screenshot 2024-08-05 at 18 34 15

Both action buttons as optional

Play icon is always a button that expects the href to be passed. Can run commands or open tabs

Info icon can have href property and title property. These properties are independent.
If the button has a href, then it can work as the play button. If the button has a title, it renders the tooltip on hover

Screenshot 2024-07-30 at 16 35 14

Onboarding items state is based on UsageTracker. Some items check if Views or webViews are shown, other checks if commands are executed, and there are 2 specific states:
- connect with repo host is based on current connection state
- codelens has its' own track key and it's tracked on the first codelens os shown

Info icons lead to corresponding youtube video part from the previous onboarding. Play buttons inspect corresponding app views.

https://github.com/user-attachments/assets/fb4887f0-92b7-43ad-9ddd-0214a3004175

# Checklist

- [x] I have followed the guidelines in the Contributing document
- [x] My changes follow the coding style of this project
- [ ] My changes build without any errors or warnings
- [x] My changes have been formatted and linted
- [x] My changes include any required corresponding changes to the documentation (including CHANGELOG.md and README.md)
- [x] My changes have been rebased and squashed to the minimal number (typically 1) of relevant commits
- [x] My changes have a descriptive commit message with a short title, including a `Fixes $XXX -` or `Closes #XXX -` prefix to auto-close the issue that your PR addresses

### CODE REVIEW

File: src/webviews/apps/shared/components/onboarding/progress-tracker.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+		if (!this.stepCount) return 0;
+		return this.progress / this.stepCount;
+	}
+
+	protected override render() {
+		return html`
+			<span>${this.progress}/${this.stepCount}</span>
+			<gl-progress-line value=${this.percentProgress} valueRate=${this.stepCount}></gl-progress-line>
+		`;
+	}
```
Comment (d13) [inline src/webviews/apps/shared/components/onboarding/progress-tracker.ts:39]: The progress element does most of this for you and is pretty easily styleable. Please look at https://developer.mozilla.org/en-US/docs/Web/HTML/Element/progress

Rough example:
```
return html`
  <label for="progress-bar">${this.progress}/${this.stepCount}</label>
  <progress id="progress-bar" value=${this.progress} max=${this.stepCount}>${this.percentProgress * 100}%</progress>
`;
```

File: src/webviews/apps/home/components/home-app.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+					infoTitle=${ifDefined(onboardingItem.infoTooltip)}
+					>${onboardingItem.title}
+				</gl-onboarding-item>
+				${repeat(onboardingItem.children!, item => item.itemId, this.renderOnboardingItem.bind(this))}
+			</gl-onboarding-item-group>
+		`;
+	}
+
+	protected override render() {
+		return html` <gl-progress-tracker progress=${this.progress} stepCount=${this.stepCount}></gl-progress-tracker>
```
Comment (d13) [inline src/webviews/apps/home/components/home-app.ts:90]: This is missing the label+count with a collapsable toggle

Screenshot 2024-07-30 at 3 03 17 PM

File: src/webviews/apps/home/home.html [hunk 3]
```diff
@@ -204,6 +204,9 @@ <h1 class="alert__title">Get Started with GitLens</h1>
 					</gl-tooltip>
 				</a>
 			</div>
+			<nav class="nav-list">
+				<gl-home-app></gl-home-app>
```
Comment (d13) [inline src/webviews/apps/home/home.html:208]: Maybe something like:

```suggestion
				<gl-onboarding><span slot="title">Get Started with GitLens</span></gl-onboarding>
```

File: src/webviews/apps/shared/components/onboarding/onboarding-item.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	checked = false;
+
+	@property({ type: String })
+	playHref = '';
+
+	@property({ type: String })
+	infoTitle = '';
+
+	@property({ type: String })
+	infoHref = '';
```
Comment (d13) [inline src/webviews/apps/shared/components/onboarding/onboarding-item.ts:62]: Camel-cased properties should have a html-compatible attribute also:

```suggestion
	@property({ type: String, attribute: 'play-href' })
	playHref = '';

	@property({ type: String, attribute: 'info-title' })
	infoTitle = '';

	@property({ type: String, attribute: 'info-href' })
	infoHref = '';
```

File: src/webviews/apps/shared/components/onboarding/onboarding-item.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+	private get infoPresented() {
+		return Boolean(this.infoHref) || Boolean(this.infoTitle);
+	}
+
+	private renderPlay() {
+		return html`<a href=${this.playHref} class="action"
+			><gl-button appearance="toolbar" ><code-icon icon="play-circle"></gl-button></a>`;
```
Comment (d13) [inline src/webviews/apps/shared/components/onboarding/onboarding-item.ts:70]: gl-button has an href attr

```suggestion
		return html`<gl-button href=${this.playHref} appearance="toolbar" ><code-icon icon="play-circle"></gl-button>`;
```

File: src/webviews/apps/shared/components/onboarding/onboarding-item.ts [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+	private renderInfo() {
+		return html`<gl-tooltip content=${this.infoTitle} .disabled=${!this.infoTitle}
+			>${when(Boolean(this.infoHref), this.renderInfoLink.bind(this), this.renderInfoIcon.bind(this))}</gl-tooltip
+		>`;
+	}
+
+	protected override render() {
+		return html`
+			<gl-onboarding-checkbox ?checked=${this.checked}> </gl-onboarding-checkbox>
```
Comment (d13) [inline src/webviews/apps/shared/components/onboarding/onboarding-item.ts:92]: Not sure this needs to be a separate component, I don't see reusing this anywhere else.

File: src/webviews/apps/shared/components/onboarding/onboarding-item.ts [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+	private renderInfoIcon() {
+		return html`<div class="info action">
+			<code-icon class=${classMap({ info: !this.infoHref })} icon="info"></code-icon>
+		</div>`;
+	}
+
+	private renderInfo() {
+		return html`<gl-tooltip content=${this.infoTitle} .disabled=${!this.infoTitle}
+			>${when(Boolean(this.infoHref), this.renderInfoLink.bind(this), this.renderInfoIcon.bind(this))}</gl-tooltip
+		>`;
```
Comment (d13) [inline src/webviews/apps/shared/components/onboarding/onboarding-item.ts:87]: gl-button has a `tooltip` attribute so the tooltip wrapper isn't needed

File: src/webviews/apps/shared/components/onboarding/progress-line.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+	@property({ type: Number })
+	valueRate: number = 20;
+
+	static override readonly styles = css`
+		.progress-container {
+			display: block;
+			height: 7px;
+			width: 100%;
+			border-radius: 4px;
+			background-color: var(--progress-background-color, gray);
```
Comment (d13) [inline src/webviews/apps/shared/components/onboarding/progress-line.ts:21]: Switch this to `--vscode-progressBar-background`

Comment (d13) [reply src/webviews/apps/shared/components/onboarding/progress-line.ts:21]: For reference, here's the main color tokens available to use:
- [provided by VSCode](https://code.visualstudio.com/api/references/theme-color)
- [provided by GitLens](https://github.com/gitkraken/vscode-gitlens/blob/16a8590f77f98070f43b05bf677e293f4867e35c/src/webviews/apps/shared/styles/theme.scss)

File: src/webviews/apps/home/components/home-app.ts [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+import { onboardingConfiguration } from '../model/gitlens-onboarding';
+
+@customElement('gl-home-app')
+export class GlHomeApp extends LitElement {
+	@property({ type: Object })
+	state?: OnboardingState;
+
+	static override readonly styles = css`
+		gl-progress-tracker {
+			padding: 0 20px;
```
Comment (d13) [inline src/webviews/apps/home/components/home-app.ts:17]: I would remove this padding and let the actual home page (home.scss) control the spacing around the component for better flexibility.

File: src/webviews/apps/home/components/home-app.ts [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+			--progress-color: var(--vscode-charts-blue);
```
Comment (d13) [inline src/webviews/apps/home/components/home-app.ts:18]: We're working to prefix all our web component custom properties with `--gl-`.

File: src/webviews/apps/home/home.ts [hunk 11]
```diff
@@ -17,12 +18,31 @@ import '../shared/components/button';
 import '../shared/components/code-icon';
 import '../shared/components/feature-badge';
 import '../shared/components/overlays/tooltip';
+import '../shared/components/onboarding/progress-line';
+import '../shared/components/onboarding/progress-tracker';
+import '../shared/components/onboarding/onboarding-item';
+import '../shared/components/onboarding/onboarding-item-group';
+import '../shared/components/onboarding/onboarding-checkbox';
```
Comment (d13) [inline src/webviews/apps/home/home.ts:25]: We should import components in the files they are being used in, so most of these import should move into the main web component.

File: src/webviews/apps/shared/components/onboarding/progress-line.ts (hunk 8 above)
Comment (nzaytsev) [reply src/webviews/apps/shared/components/onboarding/progress-line.ts:21]: Can I add these links to the confluence (initiatives section)?

File: src/webviews/apps/shared/components/onboarding/progress-tracker.ts (hunk 1 above)
Comment (nzaytsev) [reply src/webviews/apps/shared/components/onboarding/progress-tracker.ts:39]: done. I tried to run `watch & run (web)` profile to check progress styles on firefox (as styling is based on webkit properties and may be not multiplatform), but it always launch vscode desktop app. I tried to render progress with custom styles in firefox manually (using devtools inpector), it works fine, but I dind't check safari

File: src/webviews/apps/shared/components/onboarding/onboarding-item.ts (hunk 4 above)
Comment (nzaytsev) [reply src/webviews/apps/shared/components/onboarding/onboarding-item.ts:62]: done

File: src/webviews/apps/shared/components/onboarding/onboarding-item.ts (hunk 6 above)
Comment (nzaytsev) [reply src/webviews/apps/shared/components/onboarding/onboarding-item.ts:92]: fixed

File: src/webviews/apps/home/home.ts (hunk 11 above)
Comment (nzaytsev) [reply src/webviews/apps/home/home.ts:25]: done

File: src/webviews/apps/home/home.html (hunk 3 above)
Comment (nzaytsev) [reply src/webviews/apps/home/home.html:208]: ok

File: src/webviews/apps/shared/components/onboarding/onboarding-item.ts (hunk 5 above)
Comment (nzaytsev) [reply src/webviews/apps/shared/components/onboarding/onboarding-item.ts:70]: used

File: src/webviews/apps/shared/components/onboarding/onboarding-item.ts (hunk 7 above)
Comment (nzaytsev) [reply src/webviews/apps/shared/components/onboarding/onboarding-item.ts:87]: used

File: src/webviews/apps/home/components/home-app.ts [hunk 12]
```diff
@@ -0,0 +1,105 @@
+import { css, html, LitElement } from 'lit';
+import { customElement, property } from 'lit/decorators.js';
+import { ifDefined } from 'lit/directives/if-defined.js';
+import { repeat } from 'lit/directives/repeat.js';
+import { when } from 'lit/directives/when.js';
+import type { OnboardingState } from '../../../home/protocol';
+import type { OnboardingItemConfiguration } from '../model/gitlens-onboarding';
+import { onboardingConfiguration } from '../model/gitlens-onboarding';
+
+@customElement('gl-home-app')
```
[the remark this answers is not in the document]
Comment (nzaytsev) [reply src/webviews/apps/home/components/home-app.ts:10]: done

File: src/webviews/apps/home/components/home-app.ts (hunk 10 above)
Comment (nzaytsev) [reply src/webviews/apps/home/components/home-app.ts:18]: done

File: src/webviews/apps/home/components/home-app.ts (hunk 9 above)
Comment (nzaytsev) [reply src/webviews/apps/home/components/home-app.ts:17]: done

Comment (nzaytsev) [reply src/webviews/apps/home/components/home-app.ts]: I had to revert it and move the padding to the accordion to make the hovered(focused) style full-width

**Review by nzaytsev — COMMENTED** [review]
Some questions should be answered, maybe some info links and tooltips should be improved

File: src/webviews/apps/shared/components/onboarding/onboarding.ts [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+import './onboarding-item';
+import './onboarding-item-group';
+import './progress-tracker';
+import '../accordion/accordion';
+
+@customElement('gl-onboarding')
+export class GlOnboarding<
+	OnboardingState extends OnboardingStateTemplate,
+	OnboardingItem extends string,
+> extends LitElement {
```
[the remark this answers is not in the document]
Comment (nzaytsev) [reply src/webviews/apps/shared/components/onboarding/onboarding.ts:17]: done

File: src/webviews/apps/shared/components/onboarding/onboarding.ts [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+
+	@state()
+	isExpanded: boolean = true;
+
+	override connectedCallback(): void {
+		super.connectedCallback();
+		this.isExpanded = !this.finished;
+	}
+
+	static override readonly styles = css`
```
[the remark this answers is not in the document]
Comment (nzaytsev) [reply src/webviews/apps/shared/components/onboarding/onboarding.ts:18]: done

File: src/webviews/apps/shared/components/onboarding/onboarding-item.ts [hunk 15]
```diff
@@ -0,0 +1,155 @@
+import { css, html, LitElement } from 'lit';
+import { customElement, property } from 'lit/decorators.js';
+import { classMap } from 'lit/directives/class-map.js';
+import { ifDefined } from 'lit/directives/if-defined.js';
+import { when } from 'lit/directives/when.js';
+
+@customElement('gl-onboarding-item')
+export class GlOnboardingItem extends LitElement {
```
[the remark this answers is not in the document]
Comment (nzaytsev) [reply src/webviews/apps/shared/components/onboarding/onboarding-item.ts:8]: done

File: src/webviews/apps/shared/components/accordion/accordion.ts [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+ * @cssprop --gk-accordion-details-color - Defines the button color of the accordion details section.
+ * @cssprop --gk-accordion-details-padding - Defines the accordion details section padding.
+
+ * @csspart wrapper - the wrapping div around the accordion elements
+ * @csspart button - the accordion button
+ * @csspart details - the accordion details section
+ */
+@customElement('gl-accordion')
+export class Accordion extends GlElement {
+	static override readonly styles = accordionBaseStyles;
```
[the remark this answers is not in the document]
Comment (nzaytsev) [reply src/webviews/apps/shared/components/accordion/accordion.ts:36]: done

File: src/webviews/apps/shared/components/accordion/accordion.ts [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+			xmlns="http://www.w3.org/2000/svg"
+			viewBox="0 0 512 512"
+			width="16"
+			height="16"
+		>
+			<path
+				fill="currentColor"
+				d="M239 401c9.4 9.4 24.6 9.4 33.9 0L465 209c9.4-9.4 9.4-24.6 0-33.9s-24.6-9.4-33.9 0l-175 175L81 175c-9.4-9.4-24.6-9.4-33.9 0s-9.4 24.6 0 33.9L239 401z"
+			></path>
+		</svg>`;
```
[the remark this answers is not in the document]
Comment (nzaytsev) [reply src/webviews/apps/shared/components/accordion/accordion.ts]: replaced

**Comment by d13** [comment]
Converting this back to draft since the scope of this is changing quite a lot

**Review by axosoft-ramint — COMMENTED** [review]
**Edit:** commented in the wrong PR, sorry!

**Comment by nzaytsev** [comment]
looks like it's no longer needed

**Pull request closed without merging**